File: 20251220 - UI_MAP_BUILDER_INTEGRATION/topologies/simple_path.py
# src/map_generator/topologies/simple_path.py

import math
import random
import copy
from .base_topology import BaseTopology
from src.map_generator.models.path_info import PathInfo, Coord

class SimplePathTopology(BaseTopology):
    """
    (Nâng cấp) Tạo ra một con đường ngắn cho các bài tập giới thiệu.
    Có khả năng tùy chỉnh độ dài và hình dạng qua params.
    """
    def generate_path_info_variants(self, grid_size: tuple, params: dict, max_variants: int) -> "Iterator[PathInfo]":
        """
        [KIẾN TRÚC MỚI] Tạo ra các biến thể đường đi đơn giản với độ dài khác nhau.
        """
        count = 0
        # Lặp qua các độ dài có thể có
        for length in range(2, 10):
            if count >= max_variants: break
            
            variant_params = copy.deepcopy(params)
            variant_params['path_length'] = length
            # Có thể thêm logic để thay đổi cả 'turns' ở đây nếu muốn
            yield self.generate_path_info(grid_size, variant_params)
            count += 1
    
    def generate_path_info(self, grid_size: tuple, params: dict) -> PathInfo:
        print(f"    LOG: Generating 'simple_path' with params: {params}")
        
        # [CẢI TIẾN] Hỗ trợ path_length là một khoảng [min, max] để tạo sự đa dạng.
        path_length_param = params.get('path_length', [3, 5])
        if isinstance(path_length_param, list) and len(path_length_param) == 2:
            path_length = random.randint(path_length_param[0], path_length_param[1])
        else:
            path_length = int(path_length_param)

        
        # [SỬA LỖI] Đọc tham số 'turns' thay vì 'pattern' để quyết định hình dạng.
        # Nếu có 'turns' > 0, tạo đường có góc cua. Nếu không, tạo đường thẳng.
        num_turns = params.get('turns', 0)
        pattern = 'corner' if num_turns > 0 else 'straight'
        
        max_dim = min(grid_size[0], grid_size[2])
        if path_length >= max_dim - 3:
            path_length = max_dim - 4
        if path_length < 1:
            path_length = 1

        # [FIX] Tính toán không gian cần thiết cho đường đi có góc cua
        required_space = math.ceil(path_length / 2) if num_turns > 0 else path_length
        start_x = random.randint(1, grid_size[0] - required_space - 2)
        start_z = random.randint(1, grid_size[2] - required_space - 2)
        start_pos: Coord = (start_x, 0, start_z)

        path_coords: list[Coord] = []
        target_pos: Coord
        
        current_pos = list(start_pos)
        
        if pattern == 'straight':
            for _ in range(path_length):
                current_pos[0] += 1
                path_coords.append(tuple(current_pos))
            target_pos = (current_pos[0] + 1, current_pos[1], current_pos[2])

        elif pattern == 'corner':
            if path_length < 2:
                half_len, remaining_len = path_length, 0
            else:
                half_len = path_length // 2
                remaining_len = path_length - half_len
            
            for _ in range(half_len):
                current_pos[0] += 1
                path_coords.append(tuple(current_pos))
                
            for _ in range(remaining_len):
                current_pos[2] += 1
                path_coords.append(tuple(current_pos))
                
            target_pos = (current_pos[0], current_pos[1], current_pos[2] + 1)
        else:
             for _ in range(path_length):
                current_pos[0] += 1
                path_coords.append(tuple(current_pos))
             target_pos = (current_pos[0] + 1, current_pos[1], current_pos[2])

        # [NEW] Tạo metadata cho Placers
        midpoint = path_coords[len(path_coords) // 2] if path_coords else start_pos
        
        metadata = {
            "topology_type": "simple_path",
            "segment": path_coords,
            "pattern": pattern,
            "path_length": path_length,
            # [NEW] Semantic positions for simple progression
            "semantic_positions": {
                'start': start_pos,
                'end': target_pos,
                'midpoint': midpoint,
                'optimal_start': 'start',
                'optimal_end': 'end',
                'valid_pairs': [
                    {
                        'name': 'basic_traversal_easy',
                        'start': 'start',
                        'end': 'end',
                        'path_type': 'full_path',
                        'strategies': ['simple_progression', 'alternating_patterns'],
                        'difficulty': 'EASY',
                        'teaching_goal': 'Basic item collection'
                    },
                    {
                        'name': 'halfway_challenge_medium',
                        'start': 'start',
                        'end': 'midpoint',
                        'path_type': 'partial_path',
                        'strategies': ['simple_progression'],
                        'difficulty': 'MEDIUM',
                        'teaching_goal': 'Partial navigation'
                    },
                    {
                        'name': 'reverse_path_hard',
                        'start': 'end',
                        'end': 'start',
                        'path_type': 'reversed',
                        'strategies': ['simple_progression', 'corner_logic'],
                        'difficulty': 'HARD',
                        'teaching_goal': 'Reverse navigation'
                    }
                ]
            }
        }

        return PathInfo(
            start_pos=start_pos,
            target_pos=target_pos,
            path_coords=path_coords,
            metadata=metadata
        )

File: packages/academic-map-generator/src/generator/topologies/SimplePath.ts
/**
 * Simple Path Topology (PORTED FROM PYTHON)
 * Creates a simple path (straight or corner) for introductory lessons.
 */

import { BaseTopology } from './BaseTopology';
import { IPathInfo, Coord } from '../types';

export class SimplePathTopology extends BaseTopology {
  *generatePathInfoVariants(
    gridSize: [number, number, number],
    params: Record<string, any>,
    maxVariants: number
  ): Generator<IPathInfo> {
    let count = 0;
    for (let length = 2; length < 10; length++) {
        if (count >= maxVariants) break;
        
        const variantParams = { ...params };
        variantParams.path_length = length;
        yield this.generatePathInfo(gridSize, variantParams);
        count++;
    }
  }

  generatePathInfo(
    gridSize: [number, number, number],
    params: Record<string, any>
  ): IPathInfo {
    console.log("    LOG: Generating 'simple_path' topology...");

    // Path Length
    let pathLength = 3;
    const lenParam = params.path_length || [3, 5];
    if (Array.isArray(lenParam)) {
        pathLength = Math.floor(Math.random() * (lenParam[1] - lenParam[0] + 1)) + lenParam[0];
    } else {
        pathLength = Number(lenParam);
    }

    // Pattern
    const numTurns = params.turns || 0;
    const pattern = numTurns > 0 ? 'corner' : 'straight';

    // Min/Max Checks
    const maxDim = Math.min(gridSize[0], gridSize[2]);
    if (pathLength >= maxDim - 3) pathLength = maxDim - 4;
    pathLength = Math.max(1, pathLength);

    // Start Pos
    const requiredSpace = numTurns > 0 ? Math.ceil(pathLength / 2) : pathLength;
    const startX = Math.floor(Math.random() * (gridSize[0] - requiredSpace - 2)) + 1;
    const startZ = Math.floor(Math.random() * (gridSize[2] - requiredSpace - 2)) + 1;
    
    // Fix Python: start_pos = (start_x, 0, start_z)
    // Python code logic ensures start position is valid.
    
    const startPos: Coord = [startX, 0, startZ];
    const pathCoords: Coord[] = [];
    let currentPos: Coord = [...startPos];
    let targetPos: Coord = [...startPos]; // placeholder

    if (pattern === 'straight') {
        for (let i = 0; i < pathLength; i++) {
            currentPos[0] += 1; // X+
            pathCoords.push([...currentPos]);
        }
        targetPos = [currentPos[0] + 1, currentPos[1], currentPos[2]];
    } else {
        // Corner
        let halfLen = 0;
        let remainingLen = 0;
        if (pathLength < 2) {
            halfLen = pathLength;
            remainingLen = 0;
        } else {
            halfLen = Math.floor(pathLength / 2);
            remainingLen = pathLength - halfLen;
        }

        for (let i = 0; i < halfLen; i++) {
            currentPos[0] += 1;
            pathCoords.push([...currentPos]);
        }

        for (let i = 0; i < remainingLen; i++) {
            currentPos[2] += 1;
            pathCoords.push([...currentPos]);
        }
        targetPos = [currentPos[0], currentPos[1], currentPos[2] + 1];
    }
    
    // Metadata
    const midpoint = pathCoords.length > 0 
        ? pathCoords[Math.floor(pathCoords.length / 2)] 
        : startPos;

    const metadata = {
        topology_type: "simple_path",
        segment: pathCoords, // Python uses 'segment'
        pattern: pattern,
        path_length: pathLength,
        semantic_positions: {
            start: startPos,
            end: targetPos,
            midpoint: midpoint,
            optimal_start: 'start',
            optimal_end: 'end',
            valid_pairs: [
                {
                    name: 'basic_traversal_easy',
                    start: 'start',
                    end: 'end',
                    path_type: 'full_path',
                    strategies: ['simple_progression', 'alternating_patterns'],
                    difficulty: 'EASY',
                    teaching_goal: 'Basic item collection'
                },
                {
                    name: 'halfway_challenge_medium',
                    start: 'start',
                    end: 'midpoint',
                    path_type: 'partial_path',
                    strategies: ['simple_progression'],
                    difficulty: 'MEDIUM',
                    teaching_goal: 'Partial navigation'
                },
                {
                    name: 'reverse_path_hard',
                    start: 'end',
                    end: 'start',
                    path_type: 'reversed',
                    strategies: ['simple_progression', 'corner_logic'],
                    difficulty: 'HARD',
                    teaching_goal: 'Reverse navigation'
                }
            ]
        }
    };

    // Deduplicate placement coords (all walkable tiles)
    const dedupedPlacement = this._deduplicateCoords(pathCoords);
    
    // Compute path_coords using BFS pathfinding
    const computedPath = this.computePathCoords(startPos, targetPos, dedupedPlacement);

    return {
        start_pos: startPos,
        target_pos: targetPos,
        path_coords: computedPath,          // DYNAMIC: shortest path
        placement_coords: dedupedPlacement, // STATIC: all walkable tiles
        obstacles: [],
        metadata: metadata
    };
  }
}
